src/extension.ts

```diff
@@ -216,16 +218,19 @@ async function startLanguageServer() {
         }
     })
 
+    if (g_watchedEnvironmentFile) {
+        unwatchFile(g_watchedEnvironmentFile)
+    }
+
     // automatic environement refreshing
-    const targetFile = (await jlpkgenv.getProjectFilePaths(jlEnvPath)).manifest_toml_path
-    const disposableWatcher = vscode.workspace.createFileSystemWatcher(targetFile) // TODO: needs to fix backslashes on Windows ?
-    g_context.subscriptions.push(
-        disposableWatcher,
-        disposableWatcher.onDidChange(_ => {
+    g_watchedEnvironmentFile = (await jlpkgenv.getProjectFilePaths(jlEnvPath)).manifest_toml_path
+    // polling watch for robustness
+    watchFile(g_watchedEnvironmentFile, { interval: 10000 }, (curr, prev) => {

```

Comment:
I'm not familiar with file systems honestly, but as far as I read the [doc of `fs.watchFile`](https://nodejs.org/api/fs.html#fs_fs_watchfile_filename_options_listener), I think using `fs.watch` is more robust and preferred way ?

> Using fs.watch() is more efficient than fs.watchFile and fs.unwatchFile. fs.watch should be used instead of fs.watchFile and fs.unwatchFile when possible.

Could you tell why the polling is more robust for this case ?
Well, there doesn't seem to be any "unwatch API" for that, I guess we can also discard the last callback if we've started to watch for a different file.

Reply:
[`fs.watch` sounded kinda scary to me](https://nodejs.org/api/fs.html#fs_caveats). I'm not sure if it really matters either way though.

> Well, there doesn't seem to be any "unwatch API" for that, I guess we can also discard the last callback if we've started to watch for a different file.

```
const watcher = fs.watch(...)
watcher.close()
```
does the trick in that case.

Reply:
yeah, I just saw the section; let's go with `fs.watchFile` then